File: packages/google_mlkit_digital_ink_recognition/android/src/main/java/com/google_mlkit_digital_ink_recognition/DigitalInkRecognizer.java
package com.google_mlkit_digital_ink_recognition;

import androidx.annotation.NonNull;

import com.google.mlkit.common.MlKitException;
import com.google.mlkit.vision.digitalink.DigitalInkRecognition;
import com.google.mlkit.vision.digitalink.DigitalInkRecognitionModel;
import com.google.mlkit.vision.digitalink.DigitalInkRecognitionModelIdentifier;
import com.google.mlkit.vision.digitalink.DigitalInkRecognizerOptions;
import com.google.mlkit.vision.digitalink.Ink;
import com.google.mlkit.vision.digitalink.RecognitionCandidate;
import com.google.mlkit.vision.digitalink.RecognitionContext;
import com.google.mlkit.vision.digitalink.RecognitionResult;
import com.google.mlkit.vision.digitalink.WritingArea;
import com.google_mlkit_commons.GenericModelManager;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class DigitalInkRecognizer implements MethodChannel.MethodCallHandler {
    private static final String START = "vision#startDigitalInkRecognizer";
    private static final String CLOSE = "vision#closeDigitalInkRecognizer";
    private static final String MANAGE = "vision#manageInkModels";

    private final Map<String, com.google.mlkit.vision.digitalink.DigitalInkRecognizer> instances = new HashMap<>();
    private final GenericModelManager genericModelManager = new GenericModelManager();

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        String method = call.method;
        switch (method) {
            case START:
                handleDetection(call, result);
                break;
            case CLOSE:
                closeDetector(call);
                break;
            case MANAGE:
                manageModel(call, result);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    private void handleDetection(MethodCall call, final MethodChannel.Result result) {
        String tag = call.argument("model");
        DigitalInkRecognitionModel model = getModel(tag, result);
        if (model == null) return;
        if (!genericModelManager.isModelDownloaded(model)) {
            result.error("Model Error", "Model has not been downloaded yet ", null);
            return;
        }

        String id = call.argument("id");
        com.google.mlkit.vision.digitalink.DigitalInkRecognizer recognizer = instances.get(id);
        if (recognizer == null) {
            recognizer = DigitalInkRecognition.getClient(DigitalInkRecognizerOptions.builder(model).build());
            instances.put(id, recognizer);
        }

        Map<String, Object> inkMap = call.argument("ink");
        List<Map<String, Object>> strokeList = (List<Map<String, Object>>) inkMap.get("strokes");
        Ink.Builder inkBuilder = Ink.builder();
        for (final Map<String, Object> strokeMap : strokeList) {
            Ink.Stroke.Builder strokeBuilder = Ink.Stroke.builder();
            List<Map<String, Object>> pointsList = (List<Map<String, Object>>) strokeMap.get("points");
            for (final Map<String, Object> point : pointsList) {
                float x = (float) (double) point.get("x");
                float y = (float) (double) point.get("y");
                Object t0 = point.get("t");
                long t;
                if (t0 instanceof Integer) {
                    t = (int) t0;
                } else {
                    t = (long) t0;   
                }
                Ink.Point strokePoint = Ink.Point.create(x, y, t);
                strokeBuilder.addPoint(strokePoint);
            }
            inkBuilder.addStroke(strokeBuilder.build());
        }
        Ink ink = inkBuilder.build();

        RecognitionContext context = null;
        Map<String, Object> contextMap = call.argument("context");
        if (contextMap != null) {
            RecognitionContext.Builder builder = RecognitionContext.builder();
            String preContext = (String) contextMap.get("preContext");
            if (preContext != null) {
                builder.setPreContext(preContext);
            } else {
                builder.setPreContext("");
            }

            Map<String, Object> writingAreaMap = (Map<String, Object>) contextMap.get("writingArea");
            if (writingAreaMap != null) {
                float width = (float) (double) writingAreaMap.get("width");
                float height = (float) (double) writingAreaMap.get("height");
                builder.setWritingArea(new WritingArea(width, height));
            }

            context = builder.build();
        }

        if (context != null) {
            recognizer.recognize(ink, context)
                    .addOnSuccessListener(recognitionResult -> process(recognitionResult, result))
                    .addOnFailureListener(e -> result.error("recognition Error", e.toString(), null));
        } else {
            recognizer.recognize(ink)
                    .addOnSuccessListener(recognitionResult -> process(recognitionResult, result))
                    .addOnFailureListener(e -> result.error("recognition Error", e.toString(), null));
        }
    }

    private void process(RecognitionResult recognitionResult, final MethodChannel.Result result) {
        List<Map<String, Object>> candidatesList = new ArrayList<>(recognitionResult.getCandidates().size());
        for (RecognitionCandidate candidate : recognitionResult.getCandidates()) {
            Map<String, Object> candidateData = new HashMap<>();
            double score = 0;
            if (candidate.getScore() != null)
                score = candidate.getScore().doubleValue();
            candidateData.put("text", candidate.getText());
            candidateData.put("score", score);
            candidatesList.add(candidateData);
        }
        result.success(candidatesList);
    }

    private void closeDetector(MethodCall call) {
        String id = call.argument("id");
        com.google.mlkit.vision.digitalink.DigitalInkRecognizer recognizer = instances.get(id);
        if (recognizer == null) return;
        recognizer.close();
        instances.remove(id);
    }

    private void manageModel(MethodCall call, final MethodChannel.Result result) {
        String tag = call.argument("model");
        DigitalInkRecognitionModel model = getModel(tag, result);
        genericModelManager.manageModel(model, call, result);
    }

    private DigitalInkRecognitionModel getModel(String tag, final MethodChannel.Result result) {
        DigitalInkRecognitionModelIdentifier modelIdentifier;
        try {
            modelIdentifier = DigitalInkRecognitionModelIdentifier.fromLanguageTag(tag);
        } catch (MlKitException e) {
            result.error("Failed to create model identifier", e.toString(), null);
            return null;
        }
        if (modelIdentifier == null) {
            result.error("Model Identifier error", "No model was found", null);
            return null;
        }
        return DigitalInkRecognitionModel.builder(modelIdentifier).build();
    }
}

File: packages/google_mlkit_digital_ink_recognition/android/src/main/kotlin/com/google_mlkit_digital_ink_recognition/DigitalInkRecognizer.kt
package com.google_mlkit_digital_ink_recognition

import com.google.mlkit.common.MlKitException
import com.google.mlkit.vision.digitalink.common.RecognitionResult
import com.google.mlkit.vision.digitalink.recognition.DigitalInkRecognition
import com.google.mlkit.vision.digitalink.recognition.DigitalInkRecognitionModel
import com.google.mlkit.vision.digitalink.recognition.DigitalInkRecognitionModelIdentifier
import com.google.mlkit.vision.digitalink.recognition.DigitalInkRecognizerOptions
import com.google.mlkit.vision.digitalink.recognition.Ink
import com.google.mlkit.vision.digitalink.recognition.RecognitionContext
import com.google.mlkit.vision.digitalink.recognition.WritingArea
import com.google_mlkit_commons.GenericModelManager
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class DigitalInkRecognizer : MethodChannel.MethodCallHandler {
    private val instances = mutableMapOf<String, com.google.mlkit.vision.digitalink.recognition.DigitalInkRecognizer>()
    private val genericModelManager = GenericModelManager()

    companion object {
        private const val START = "vision#startDigitalInkRecognizer"
        private const val CLOSE = "vision#closeDigitalInkRecognizer"
        private const val MANAGE = "vision#manageInkModels"
    }

    override fun onMethodCall(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        when (call.method) {
            START -> {
                handleDetection(call, result)
            }

            CLOSE -> {
                closeDetector(call)
            }

            MANAGE -> {
                manageModel(call, result)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    private fun handleDetection(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val tag = call.argument<String>("model")
        val model = getModel(tag!!, result) ?: return

        genericModelManager.isModelDownloaded(
            model,
            object : GenericModelManager.CheckModelIsDownloadedCallback {
                override fun onCheckResult(isDownloaded: Boolean?) {
                    if (isDownloaded == false) {
                        result.error("Model Error", "Model has not been downloaded yet", null)
                        return
                    }
                    handleInkDetectionIfModelDownloaded(call, result, model)
                }

                override fun onError(e: Exception) {
                    result.error("Model download check failed", e.toString(), e)
                }
            },
        )
    }

    private fun handleInkDetectionIfModelDownloaded(
        call: MethodCall,
        result: MethodChannel.Result,
        model: DigitalInkRecognitionModel,
    ) {
        val id = call.argument<String>("id")!!
        val recognizer =
            instances.getOrPut(id) {
                DigitalInkRecognition.getClient(DigitalInkRecognizerOptions.builder(model).build())
            }

        val inkMap = call.argument<Map<String, Any>>("ink")!!
        val strokeList = inkMap["strokes"] as List<Map<String, Any>>
        val inkBuilder = Ink.builder()

        for (strokeMap in strokeList) {
            val strokeBuilder = Ink.Stroke.builder()
            val pointsList = strokeMap["points"] as List<Map<String, Any>>
            for (point in pointsList) {
                val x = (point["x"] as Double).toFloat()
                val y = (point["y"] as Double).toFloat()
                val t =
                    when (val t0 = point["t"]) {
                        is Int -> t0.toLong()
                        else -> t0 as Long
                    }

                strokeBuilder.addPoint(Ink.Point.create(x, y, t))
            }
            inkBuilder.addStroke(strokeBuilder.build())
        }

        val ink = inkBuilder.build()

        val contextMap = call.argument<Map<String, Any>>("context")
        val context =
            contextMap?.let {
                val builder = RecognitionContext.builder()
                builder.setPreContext((it["preContext"] as? String) ?: "")

                (it["writingArea"] as? Map<*, *>)?.let { areaMap ->
                    val width = (areaMap["width"] as Double).toFloat()
                    val height = (areaMap["height"] as Double).toFloat()
                    builder.setWritingArea(WritingArea(width, height))
                }
                builder.build()
            }

        val onSuccess = { recognitionResult: RecognitionResult -> process(recognitionResult, result) }
        val onFailure = { e: Exception -> result.error("recognition Error", e.toString(), null) }

        if (context != null) {
            recognizer
                .recognize(ink, context)
                .addOnSuccessListener(onSuccess)
                .addOnFailureListener(onFailure)
        } else {
            recognizer
                .recognize(ink)
                .addOnSuccessListener(onSuccess)
                .addOnFailureListener(onFailure)
        }
    }

    private fun process(
        recognitionResult: RecognitionResult,
        result: MethodChannel.Result,
    ) {
        val candidateList =
            recognitionResult.candidates.map { candidate ->
                mapOf(
                    "text" to candidate.text,
                    "score" to (candidate.score?.toDouble() ?: 0.0),
                )
            }
        result.success(candidateList)
    }

    private fun getModel(
        tag: String,
        result: MethodChannel.Result,
    ): DigitalInkRecognitionModel? {
        val modelIdentifier =
            try {
                DigitalInkRecognitionModelIdentifier.fromLanguageTag(tag)
            } catch (e: MlKitException) {
                result.error("Failed to create model identifier", e.toString(), null)
                return null
            }

        if (modelIdentifier == null) {
            result.error("Model Identifier error", "No model was found", null)
            return null
        }

        return DigitalInkRecognitionModel.builder(modelIdentifier).build()
    }

    private fun manageModel(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
    }

    private fun closeDetector(call: MethodCall) {
        val id = call.argument<String>("id") ?: return
        instances.remove(id)?.close()
    }
}
